src/test/scala/com/fulcrumgenomics/bam/AddReadGroupsByNameTest.scala

```diff
@@ -0,0 +1,123 @@
+    val out = getAddReadGroupsByNameOutput(
+      "instrument:run-number:flowcell-id:1:2:3:4",
+      "instrument:run-number:flowcell-id:1:2:3:4",
+      "instrument:run-number:flowcell-id:1:3:3:4",
+      "instrument:run-number:flowcell-id:1:2:4:4",
+      "instrument:run-number:flowcell-id:1:2:3:5"
+    )
+    val readGroups = getHeader(out).getReadGroups
+    readGroups should have size 1
+    readGroups.map(_.getId) should contain theSameElementsAs Seq("1")
+    readGroups.map(_.getPlatformUnit) should contain theSameElementsAs Seq("flowcell-id.1")
+  }
+
+  it should "add a two read groups for reads from two lanes in one flowcell" in {
+    val out = getAddReadGroupsByNameOutput(
+      "instrument:run-number:flowcell-id:1:2:3:4",
+      "instrument:run-number:flowcell-id:2:2:3:4"
+    )
+    val readGroups = getHeader(out).getReadGroups
+    readGroups should have size 2
+    readGroups.map(_.getId) should contain theSameElementsAs Seq("1", "2")
+    readGroups.map(_.getPlatformUnit) should contain theSameElementsAs Seq("flowcell-id.1", "flowcell-id.2")
+  }
+
+  it should "add two read groups for reads from the first lane for two flowcells" in {
+    val out = getAddReadGroupsByNameOutput(
+      "instrument:run-number:flowcell-id-1:1:2:3:4",
+      "instrument:run-number:flowcell-id-2:1:2:3:4"
+    )
+    val readGroups = getHeader(out).getReadGroups
+    readGroups should have size 2
+    readGroups.map(_.getId) should contain theSameElementsAs Seq("1", "2")
+    readGroups.map(_.getPlatformUnit) should contain theSameElementsAs Seq("flowcell-id-1.1", "flowcell-id-2.1")
+  }
+
+  it should "add two read groups for reads from two different instruments" in {
+    val out = getAddReadGroupsByNameOutput(
+      "instrument-1:run-number:flowcell-id:1:2:3:4",
+      "instrument-2:run-number:flowcell-id:1:2:3:4"
+    )
+    val readGroups = getHeader(out).getReadGroups
+    readGroups should have size 2
+    readGroups.map(_.getId) should contain theSameElementsAs Seq("1", "2")
+    readGroups.map(_.getPlatformUnit) should contain theSameElementsAs Seq("flowcell-id.1", "flowcell-id.1")
+    readGroups.map(_.getPlatformModel) should contain theSameElementsAs Seq("instrument-1.run-number", "instrument-2.run-number")
+  }
+
+  it should "add two read groups for reads from two different run numbers" in {
+    val out = getAddReadGroupsByNameOutput(
+      "instrument:run-number-1:flowcell-id:1:2:3:4",
+      "instrument:run-number-2:flowcell-id:1:2:3:4"
+    )
+    val readGroups = getHeader(out).getReadGroups
+    readGroups should have size 2
+    readGroups.map(_.getId) should contain theSameElementsAs Seq("1", "2")
+    readGroups.map(_.getPlatformUnit) should contain theSameElementsAs Seq("flowcell-id.1", "flowcell-id.1")
+    readGroups.map(_.getPlatformModel) should contain theSameElementsAs Seq("instrument.run-number-1", "instrument.run-number-2")
+  }
+}
```

Comment:
I think one more test that ensures that if you pass all the optional parameters in that they show up in the RG headers would be good.